### PR TITLE
Add support for functions

### PR DESCRIPTION
**Pull request by dshcherb**
Add support for functions

### CODE REVIEW

**Comment by dshcherb** [comment]
Some design considerations related to `FunctionEvent` deferring:

* Ban deferring of functions to avoid the situation where function hook tools cannot be used as the context is gone;
* Keep the regular mechanism of deferring for `FunctionEvent`s but a reference to a function and its parameters may be gone by the time you handle the event (much like with `RelationEvent`s and relation data when a relation is gone by the time of access). function-{get, set, fail} will not be usable for deferred functions (functions will succeed after being deferred because they are one-time executions);
* Keep the regular mechanism of deferring but persist function parameters along with FunctionEvent. Hook tools will not work but getting function parameters will work because of the persistence layer. This can be viewed as a mechanism to submit jobs *with delayed completion* (possibly spanning multiple hook executions) where the results are only viewable via the charm log and charm status.

File: ops/main.py [hunk 1]
```diff
@@ -105,22 +121,21 @@ def _get_event_args(charm, bound_event):
...
-            remote_app_name = remote_unit_name.split('/')[0]
-        args = [relation, model.get_app(remote_app_name)]
-        if remote_unit_name:
-            args.append(model.get_unit(remote_unit_name))
-        return args, {}
-    elif relation:
-        return [relation], {}
+        remote_app_name = os.environ.get('JUJU_REMOTE_APP', '')
+        remote_unit_name = os.environ.get('JUJU_REMOTE_UNIT', '')
+        if remote_app_name or remote_unit_name:
```
Comment (niemeyer) [inline ops/main.py]: This logic changed shape in a way unrelated to functions. The original logic seemed fine, and in either case it seems unrelated to functions so would be better done separately.

File: ops/model.py [hunk 2]
```diff
@@ -622,3 +667,21 @@ def storage_add(self, name, count=1):
...
         self._run('storage-add', f'{name}={count}')
+
+    def function_get(self):
+        command_name = 'function-get'
+        if not shutil.which(command_name):
+            command_name = 'action-get'
+        return self._run(command_name, return_output=True, use_json=True)
+
+    def function_set(self, **kwargs):
+        command_name = 'function-set'
```
Comment (jameinel) [inline ops/model.py]: Is it worth caching the command name rather than doing a $PATH lookup on every call to function_set() ? It doesn't change for the lifetime of a process, and the LazyMapping means that every call to "result.foo = X" is going to trigger this.

File: test/test_model.py [hunk 3]
```diff
@@ -758,6 +843,52 @@ def test_storage_tool_errors(self):
...
+            lambda: fake_script(self, 'function-fail', f'echo fooerror >&2 ; exit 1'),
+            lambda: self.backend.function_fail('fail-message'),
+            ops.model.ModelError,
+            [["function-fail", "'fail-message'"]],
+        )]
+        for do_fake, run, exception, calls in test_cases:
+            do_fake()
+            with self.assertRaises(exception):
+                run()
+            self.assertEqual(fake_script_calls(self, clear=True), calls)
```
Comment (niemeyer) [inline test/test_model.py]: Given this loop, which consists almost exclusively of calling the individual lambdas right above, it looks like the "table" style is damaging readability here instead of helping. Also, this is leaving the prior faked scripts behind, which is worse than not doing that.

I suspect having actual independent test cases here will be both cleaner and more readable.

File: test/test_model.py [hunk 4]
```diff
@@ -758,6 +843,52 @@ def test_storage_tool_errors(self):
...
+            lambda: fake_script(self, 'action-fail', f'echo fooerror >&2 ; exit 1'),
+            lambda: self.backend.function_fail('fail-message'),
+            ops.model.ModelError,
+            [["action-fail", "'fail-message'"]],
+        )]
+        for do_fake, run, exception, calls in test_cases:
+            do_fake()
+            with self.assertRaises(exception):
+                run()
+            self.assertEqual(fake_script_calls(self, clear=True), calls)
```
Comment (niemeyer) [inline test/test_model.py]: Same.

File: ops/model.py [hunk 5]
```diff
@@ -476,6 +480,45 @@ def location(self):
...
+class FunctionParameters(LazyMapping):
+
+    def __init__(self, backend):
+        self._backend = backend
+
+    def _load(self):
+        return self._backend.function_get()
+
+
+class FunctionResults(LazyMapping, MutableMapping):
```
Comment (niemeyer) [inline ops/model.py]: As we discussed in our call today, this doesn't feel right for functions. The context of the function is only valid while the event is being handled. Once the event is over, juju will consider the function as terminated and these parameters and results are meaningless.

This would probably be better modeled as a richer event type.

File: ops/charm.py [hunk 6]
```diff
@@ -239,3 +269,15 @@ class PayloadMeta:
...
+
+class FunctionMeta:
+    """An object containing metadata about an function"""
+
+    def __init__(self, name, raw=None):
+        raw = raw or {}
+        self.function_name = name
+        self.title = raw.get('title', '')
+        self.description = raw.get('description', '')
+        self.parameters = raw.get('params', {})  # <parameter name>: <JSON-schema>
```
[the remark this answers is not in the document]
Comment (johnsca) [reply ops/charm.py]: It actually is a data structure conforming to the [JSON Schema specification](https://json-schema.org/specification.html) ([per docs](https://discourse.jujucharms.com/t/actions-for-the-charm-author/1113#heading--options-and-format-actionsyaml)).

Maybe something like this would be more descriptive:

```python
        self.parameters = raw.get('params', {})  # Mapping of names to JSON Schema definitions
```

File: ops/main.py [hunk 7]
```diff
@@ -30,6 +30,13 @@ def _get_charm_dir():
 def _load_metadata(charm_dir):
     with open(charm_dir / 'metadata.yaml') as f:
         metadata = yaml.load(f, Loader=yaml.SafeLoader)
+
+    functions_meta = charm_dir / 'functions.yaml'
+    if not functions_meta.exists():
+        functions_meta = charm_dir / 'actions.yaml'
```
[the remark this answers is not in the document]
Comment (johnsca) [reply ops/main.py]: Juju doesn't tell us if the newer functions API is supported and we can't tell from the version, since it's behind a feature flag.  However, from testing with the feature flag enabled, I've found:

* If `functions.yaml` is defined, both `actions.yaml` and `actions/` will be entirely ignored.
* If `actions.yaml` is defined `functions/` can be used but it must contain all of the action implementations, as `actions/` will be entirely ignored; otherwise, `actions/` will be used as per current behavior.

Of course, if the feature flag is not enabled or an older Juju is used, `functions.yaml` and `functions/` will be entirely ignored.

[the remark this answers is not in the document]
Comment (johnsca) [reply ops/main.py]: After having more discussion with Ian, the intention is for the new API to always be available in 2.7+ regardless of feature flag (the flag just controls the CLI access to the feature).  Still, the actual use of the feature is driven by whether the dev uses `actions.yaml` or `functions.yaml`, and since the use of `actions/` vs `functions/` is handled by the framework, we should probably detect which filename they use, preferring `functions.yaml` like Juju does, and drive the symlinks based on that.

[the remark this answers is not in the document]
Comment (johnsca) [reply ops/main.py]: For this specific block, what about this:

```python
    for filename in ('functions.yaml', 'actions.yaml'):
        functions_meta = charm_dir / filename
        if functions_meta.exists():
            with open(functions_meta) as f:
                metadata['functions'] = yaml.load(f, Loader=yaml.SafeLoader)
            break
```

[the remark this answers is not in the document]
Comment (niemeyer) [reply ops/main.py]: This is still based on whether the files exist. A file existing or not is defined solely by the charm.. if we deploy a charm in an old juju, the existence of functions.yaml doesn't mean anything. The ideal scenario is defining whether to support it or not based on juju, not on charm data that will exist independently.

[the remark this answers is not in the document]
Comment (niemeyer) [reply ops/main.py]: Just had a call with John, and we agreed that the easiest way out is to forbid both files, functions.yaml and actions.yaml, to exist at once in the charm. That's trivial to do and effective, and prevents bad charms anyway, as that's a bad idea. :+1:

[the remark this answers is not in the document]
Comment (johnsca) [reply ops/main.py]: Per discussion, we want to consider having both styles to be a programmer error.

File: ops/main.py [hunk 8]
```diff
@@ -32,29 +32,49 @@ def _load_metadata(charm_dir):
...
+        metadata['functions_type'] = 'actions'
+        metadata['functions'] = yaml.safe_load(actions_meta.read_text())
     return metadata
 
 
-def _create_event_link(charm_dir, event_dir, target_path, bound_event):
+def _create_event_link(charm, bound_event):
     """Create a symlink for a particular event.
 
     charm_dir -- A root directory of the charm
```
Comment (dshcherb) [inline ops/main.py]: Need to fix the argument name here as well: `charm_dir` -> `charm`.

File: ops/charm.py [hunk 9]
```diff
@@ -7,19 +7,25 @@ class HookEvent(EventBase):
...
     def defer(self):
         raise RuntimeError('cannot defer function events')
 
     def restore(self, snapshot):
-        self.function = self.framework.model.function
+        event_function_name = self.handle.kind[:-len('_function')].replace('_', '-')
+        if event_function_name != self.framework.model._backend.function_name:
+            # This could only happen if the dev manually emits the function, or from a bug.
+            raise RuntimeError('function event kind does not match current function')
+        # Parms are loaded at restore rather than __init__ because the model is not available in __init__.
```
Comment (dshcherb) [inline ops/charm.py]: `s/Parms/{Params or Parameters}`

**Comment by dshcherb** [comment]
Another design consideration is the fact that for K8s charms Juju executes functions in unit pods instead of the operator pod:

https://github.com/juju/juju/blob/juju-2.7.0/worker/uniter/runner/runner.go#L288-L292

There is no guarantee that `python3` will even be in the container image of the target workload pod. Likewise, the storage for a charm directory where we currently store unit state will be different for operator and workload pods.

File: ops/main.py [hunk 10]
```diff
@@ -30,24 +30,48 @@ def _get_charm_dir():
 def _load_metadata(charm_dir):
     with open(charm_dir / 'metadata.yaml') as f:
         metadata = yaml.load(f, Loader=yaml.SafeLoader)
+
+    functions_meta = charm_dir / 'functions.yaml'
+    functions_dir = charm_dir / 'functions'
+    actions_meta = charm_dir / 'actions.yaml'
+    actions_dir = charm_dir / 'actions'
+    if functions_meta.exists():
+        if 'min-juju-version' not in metadata:
```
Comment (dshcherb) [inline ops/main.py]: I think this will end up requiring everybody to know about the `min-juju-version` requirement for functions long past 2.6.x will go EOL.

Maybe we could do a juju version check via `JUJU_VERSION` environment variable and log a warning here instead.

Comment (dshcherb) [reply ops/main.py]: We will need something like https://github.com/canonical/operator/pull/100 in order to do version Juju checking and comparison.

I suggest just dropping this check for now.

File: ops/main.py [hunk 11]
```diff
@@ -30,24 +30,43 @@ def _get_charm_dir():
...
+    actions_meta = charm_dir / 'actions.yaml'
+    if functions_meta.exists():
+        # TODO: add a version check and a warning here to make sure functions are not skipped without notice pre-2.7.0 Juju versions.
+        if actions_meta.exists():
+            raise RuntimeError('charm must not mix functions and actions metadata')
+        metadata['functions_type'] = 'functions'
+        metadata['functions'] = yaml.safe_load(functions_meta.read_text())
+    elif actions_meta.exists():
+        metadata['functions_type'] = 'actions'
+        metadata['functions'] = yaml.safe_load(actions_meta.read_text())
```
Comment (niemeyer) [inline ops/main.py]: We should try to avoid that kind of mixing of data that comes from a user file with data that we inject because we were lazy and didn't want to change the code to pass data around properly. Imagine what happens now if neither functions.yaml or actions.yaml exists, but the user puts a key labeled "functions" in metadata.yaml. Oops. It also reads improperly/confusingly, per earlier not in the PR.. it's the "raw data", except for things we stashed on it for convenience.

File: ops/main.py [hunk 12]
```diff
@@ -30,24 +30,43 @@ def _get_charm_dir():
...
     """
-    # TODO: Handle function/action events here.
-    if not issubclass(bound_event.event_type, ops.charm.HookEvent):
+    if issubclass(bound_event.event_type, ops.charm.HookEvent):
+        event_dir = charm.framework.charm_dir / 'hooks'
+        event_path = event_dir / bound_event.event_kind.replace('_', '-')
+    elif issubclass(bound_event.event_type, ops.charm.FunctionEvent):
+        event_dir = charm.framework.charm_dir / charm.framework.meta.functions_type
+        # The event_kind is suffixed with "_function" while the executable is not.
+        event_path = event_dir / bound_event.event_kind[:-len('_function')].replace('_', '-')
```
Comment (niemeyer) [inline ops/main.py]: We need something like this here, to avoid very weird bugs:
```
if not bound_event.event_kind.endswith("_function"):
        raise RuntimeError("function event name {event_name} needs _function suffix")
```

File: ops/model.py [hunk 13]
```diff
@@ -507,7 +507,13 @@ class ModelBackend:
     def __init__(self):
         self.unit_name = os.environ['JUJU_UNIT_NAME']
         self.app_name = self.unit_name.split('/')[0]
+        self.function_name = os.environ.get('JUJU_FUNCTION_NAME',
```
Comment (niemeyer) [inline ops/model.py]: This seems awkward. It breaks the idea that the model and the backend may be used for multiple events. Let's please try to keep the awkwardness of function handling into the function logic itself somehow. A cheap way for now is moving the environment checking into the function event itself. This is just a sanity check either way as far as I understand it, so shouldn't be a problem.

File: ops/model.py [hunk 14]
```diff
@@ -622,3 +628,15 @@ def storage_add(self, name, count=1):
         if not isinstance(count, int) or isinstance(count, bool):
             raise TypeError(f'storage count must be integer, got: {count} ({type(count)})')
         self._run('storage-add', f'{name}={count}')
+
+    def function_get(self):
+        return self._run(f'{self._function_cmd_prefix}-get', return_output=True, use_json=True)
+
+    def function_set(self, results):
+        self._run(f'{self._function_cmd_prefix}-set', *[f"{k}='{v}'" for k, v in results.items()])
```
Comment (niemeyer) [inline ops/model.py]: What happens if there are quotes in those values? Ideally function-set can take parameters in the 'foo=bar' style, unquoted, and ideally _run is indeed passing argv arguments into it instead of using the shell. This would indicate the quotes here are unnecessary which is the easiest fix. If either of those assumptions are wrong, this needs a more complex fix.

File: ops/model.py [hunk 15]
```diff
@@ -622,3 +628,15 @@ def storage_add(self, name, count=1):
...
+
+    def function_log(self, message):
+        self._run(f'{self._function_cmd_prefix}-log', f"'{message}'")
+
+    def function_fail(self, message=''):
+        self._run(f'{self._function_cmd_prefix}-fail', f"'{message}'")
```
Comment (niemeyer) [inline ops/model.py]: Similar note about quoting.

File: ops/charm.py [hunk 16]
```diff
@@ -190,6 +217,8 @@ def __init__(self, raw=None):
         self.payloads = {name: PayloadMeta(name, payload)
                          for name, payload in raw.get('payloads', {}).items()}
         self.extra_bindings = raw.get('extra-bindings', [])
+        self.functions_type = raw.get('functions_type', None)
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/charm.py]: This was added as a replacement for the logic I had before that passed `event_dir` to `_create_event_link`:
https://github.com/canonical/operator/pull/79/commits/b4245651e1f4db2c47fd4d6f4681f29f22f8b3a0#diff-2a757bffeaf9ecf9b797cba4ea3223bbL83-L90

It would be good to discuss if I should get back to the old logic or come up with some other alternative.

File: ops/model.py (hunk 14 above)
Comment (dshcherb) [reply ops/model.py]: You're right, this definitely needs changing as we're not passing a command to a shell.

I looked at the function-set code again and it uses `strings.SplitN(arg, "=", 2)` so the key is separated from the value based on the first equal sign.

```
juju show-action-output 2              
# ...
results:
  ran: many=equal=signs=in=the=value
status: completed
# ...
```

Thanks for noticing this.

File: ops/model.py (hunk 13 above)
Comment (dshcherb) [reply ops/model.py]: Yes, just a sanity check. I'll move it as you described.

File: ops/main.py (hunk 12 above)
Comment (dshcherb) [reply ops/main.py]: Good point.

File: ops/charm.py [hunk 17]
```diff
@@ -239,3 +268,13 @@ class PayloadMeta:
     def __init__(self, name, raw):
         self.payload_name = name
         self.type = raw['type']
+
+
+class FunctionMeta:
+    def __init__(self, name, raw=None):
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/charm.py]: We did, sorry about this one.

File: ops/charm.py [hunk 18]
```diff
@@ -239,3 +268,13 @@ class PayloadMeta:
...
+        raw = raw or {}
+        self.function_name = name
+        self.title = raw.get('title', '')
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/charm.py]: Changed to `self.name`.

File: ops/main.py [hunk 19]
```diff
@@ -30,24 +30,43 @@ def _get_charm_dir():
 def _load_metadata(charm_dir):
     with open(charm_dir / 'metadata.yaml') as f:
         metadata = yaml.load(f, Loader=yaml.SafeLoader)
+
+    functions_meta = charm_dir / 'functions.yaml'
+    actions_meta = charm_dir / 'actions.yaml'
+    if functions_meta.exists():
+        # TODO: add a version check and a warning here to make sure functions are not skipped without notice pre-2.7.0 Juju versions.
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply ops/main.py]: Dropped.
